File: modules/kotlin-native-maven-plugin/src/main/java/com/mscg/kotlinNative/maven/plugin/KotlinNativeMojo.java
package com.mscg.kotlinNative.maven.plugin;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.apache.maven.plugins.annotations.ResolutionScope;
import org.codehaus.mojo.exec.ExecMojo;
import org.codehaus.plexus.util.DirectoryScanner;

@Mojo( name = "exec", threadSafe = true, requiresDependencyResolution = ResolutionScope.TEST )
public class KotlinNativeMojo extends ExecMojo {

  /**
   * List of fileset patterns to specify source files to include in compilation.
   */
  @Parameter
  private String[] includes;

  /**
   * List of fileset patterns to specify source files to exclude from compilation.
   */
  @Parameter
  private String[] excludes;

  /**
   * The folder in which the native executable will be placed. Defaults to <code>${project.build.directory}</code>
   */
  @Parameter( property = "kotlin.compiledFolder", defaultValue = "${project.build.directory}" )
  private String compiledFolder;

  /**
   * The base name of native executable that will be generated. Defaults to <code>${project.artifactId}</code>
   */
  @Parameter( property = "kotlin.compiledFile", defaultValue = "${project.artifactId}" )
  private String compiledFile;

  /**
   * A switch to enable optimizations in the generated executables. Defaults to <code>false</code>
   */
  @Parameter( property = "kotlin.optimizations", defaultValue = "false" )
  private boolean optimizations;

  private FieldAccessor<String> _executable;

  private FieldAccessor<List<?>> _arguments;

  private void initFields() throws MojoExecutionException {
    try {
      Field field = ExecMojo.class.getDeclaredField("executable");
      field.setAccessible(true);
      _executable = new FieldAccessor<>(this, field);
    }
    catch (Exception e) {
      getLog().error(e);
      throw new MojoExecutionException("Failed to get executable property", e);
    }

    try {
      Field field = ExecMojo.class.getDeclaredField("arguments");
      field.setAccessible(true);
      _arguments = new FieldAccessor<>(this, field);
    }
    catch (Exception e) {
      getLog().error(e);
      throw new MojoExecutionException("Failed to get arguments property", e);
    }
  }

  @Override
  @SuppressWarnings("unchecked")
  public void execute() throws MojoExecutionException {
    try {
      initFields();

      if (_executable.get() == null) {
        getLog().info("Setting default value for kotlinc-native compiler");
        _executable.set("kotlinc-native");
      }

      List<Object> _arguments = (List<Object>) this._arguments.get();
      if (_arguments == null) {
        _arguments = new ArrayList<>();
        this._arguments.set(_arguments);
      }

      if (optimizations) {
        _arguments.add("-opt");
      }

      File outputFile = new File(compiledFolder, this.compiledFile);
      outputFile = outputFile.getCanonicalFile();
      String path = safeFileName(outputFile.getPath());
      getLog().info("Compiling application into basename " + path);
      _arguments.add("-o");
      _arguments.add(path);

      getLog().debug("Listing sources with include filters " + Optional.ofNullable(includes).map(Arrays::toString).orElse("<empty>") +
          " and exclude filters " + Optional.ofNullable(excludes).map(Arrays::toString).orElse("<empty>"));
      Stream<String> sourceRootsStream = project.getCompileSourceRoots().stream();
      List<String> sources = sourceRootsStream
        .map(File::new)
        .filter(File::exists)
        .flatMap(basedir -> {
          DirectoryScanner scanner = new DirectoryScanner();
          scanner.setIncludes(includes);
          scanner.setExcludes(excludes);
          scanner.setBasedir(basedir);
          scanner.scan();

          return Arrays.stream(scanner.getIncludedFiles())
              .map(file -> new File(basedir, file));
        })
        .map(t -> {
          try {
            return t.getCanonicalPath();
          }
          catch (IOException e) {
            throw new RuntimeException(e);
          }
        })
        .map(this::safeFileName)
        .collect(Collectors.toList());
      getLog().info("Adding " + sources.size() + " files to command line arguments");
      _arguments.addAll(sources);

    }
    catch (MojoExecutionException e) {
      throw e;
    }
    catch (RuntimeException e) {
      Throwable cause = e.getCause();
      throw new MojoExecutionException("An error occurred while initializing plugin configuration: " + e.getMessage(), cause);
    }
    catch (Exception e) {
      throw new MojoExecutionException("An error occurred while initializing plugin configuration: " + e.getMessage(), e);
    }

    super.execute();
  }

  private String safeFileName(String baseFileName) {
    if (baseFileName.contains(" ")) {
      return "\"" + baseFileName + "\"";
    }

    return baseFileName;
  }

  private static class FieldAccessor<T> implements Supplier<T>, Consumer<T> {

    private Object target;
    private Field field;

    public FieldAccessor(Object target, Field field) {
      this.target = target;
      this.field = field;
    }

    @Override
    @SuppressWarnings("unchecked")
    public T get() {
      try {
        return (T) field.get(target);
      }
      catch (Exception e) {
        throw new RuntimeException(e);
      }
    }

    public void set(T t) {
      try {
        field.set(target, t);
      }
      catch (Exception e) {
        throw new RuntimeException(e);
      }
    }

    @Override
    public void accept(T t) {
      set(t);
    }
  }

}

File: modules/kotlin-native-maven-plugin/src/main/kotlin/com/mscg/kotlinNative/maven/plugin/KotlinNativeMojo.kt
package com.mscg.kotlinNative.maven.plugin

import org.apache.maven.plugin.MojoExecutionException
import org.apache.maven.plugins.annotations.Mojo
import org.apache.maven.plugins.annotations.Parameter
import org.apache.maven.plugins.annotations.ResolutionScope
import org.codehaus.mojo.exec.ExecMojo
import java.lang.reflect.Field
import java.util.function.Consumer
import java.util.function.Supplier
import java.io.File
import org.codehaus.plexus.util.DirectoryScanner

@Mojo(name = "exec", threadSafe = true, requiresDependencyResolution = ResolutionScope.TEST)
class KotlinNativeMojo : ExecMojo() {
    /**
     * List of fileset patterns to specify source files to include in compilation.
     */
    @Parameter
    private var includes: Array<String>? = null

    /**
     * List of fileset patterns to specify source files to exclude from compilation.
     */
    @Parameter
    private var excludes: Array<String>? = null
    
    /**
     * The source directories containing the sources to be compiled.
     */
    @Parameter
    private var sourceDirs: List<String>? = null
        get() {
            val dirs: MutableList<String> = mutableListOf()
            val size = (field?.size ?: 0);
            if (size != 0) {
                dirs += field!!
            }
            else {
                dirs += project.compileSourceRoots.map { it as String }
            }
            dirs += (additionalSourceDirs ?: listOf())
            return dirs
        }
    
    /**
     * Additional source directories containing the sources to be compiled.
     */
    @Parameter
    private var additionalSourceDirs: List<String>? = null

    /**
     * The folder in which the native executable will be placed. Defaults to <code>${project.build.directory}</code>
     */
    @Parameter(property = "kotlin.compiledFolder", defaultValue = "\${project.build.directory}")
    private lateinit var compiledFolder: String;

    /**
     * The base name of native executable that will be generated. Defaults to <code>${project.artifactId}</code>
     */
    @Parameter(property = "kotlin.compiledFile", defaultValue = "\${project.artifactId}")
    private lateinit var compiledFile: String;

    /**
     * A switch to enable optimizations in the generated executables. Defaults to <code>false</code>
     */
    @Parameter(property = "kotlin.optimizations", defaultValue = "false")
    private var optimizations: Boolean = false;
        
    /**
     * Enable multiplatform features
     */
    @Parameter(property = "kotlin.multiPlatform", defaultValue = "false")
    private var multiPlatform: Boolean = false

    private lateinit var _executable: FieldAccessor<String>;

    private lateinit var _arguments: FieldAccessor<MutableList<*>>;

    @Throws(MojoExecutionException::class)
    private fun initFields() {
        try {
            val field = ExecMojo::class.java.getDeclaredField("executable");
            field.setAccessible(true);
            _executable = FieldAccessor(this, field);
        } catch (e: Exception) {
            log.error(e);
            throw MojoExecutionException("Failed to get executable property", e);
        }

        try {
            val field = ExecMojo::class.java.getDeclaredField("arguments");
            field.setAccessible(true);
            _arguments = FieldAccessor(this, field);
        } catch (e: Exception) {
            log.error(e);
            throw MojoExecutionException("Failed to get arguments property", e);
        }
    }
    
    @Suppress("UNCHECKED_CAST")
    @Throws(MojoExecutionException::class)
    override fun execute() {
        try {
            initFields()

            if (_executable.get() == null) {
                log.info("Setting default value for kotlinc-native compiler");
                _executable.set("kotlinc-native");
            }

            var _arguments: MutableList<Any?>? = this._arguments.get() as MutableList<Any?>?
            if (_arguments == null) {
                _arguments = mutableListOf()
                this._arguments.set(_arguments);
            }

            with(_arguments) {
                if (optimizations) {
                    add("-opt");
                }
                
                if (multiPlatform) {
                    add("-Xmulti-platform")
                }

                val targetFolder = File(compiledFolder)
                if (!targetFolder.exists()) {
                    log.info("Creating target folder ${targetFolder.canonicalPath}")
                    if (!targetFolder.mkdirs()) {
                        throw MojoExecutionException("Failed to create target folder ${targetFolder.canonicalPath} for native executable")
                    }
                }
                if (targetFolder.exists() && !targetFolder.isDirectory) {
                    throw MojoExecutionException("Path ${targetFolder.canonicalPath} is not a folder that can contain the native executable")
                }
                val outputFile = File(compiledFolder, compiledFile).getCanonicalFile();
                val path = outputFile.getPath().safeFileName();
                log.info("Compiling application into basename $path");
                add("-o");
                add(path);
                
                if (log.isDebugEnabled) {
                  log.debug("Listing sources with include filters ${includes?.contentToString() ?: "<empty>"} and exclude filters ${excludes?.contentToString() ?: "<empty>"}")                    
                }
                val sources: List<String> = (sourceDirs ?: listOf()).asSequence()
                    .map { File(it) }
                    .filter { it.exists() }
                    .flatMap { sourceBasedir ->
                        with(DirectoryScanner()) {
                            setIncludes(this@KotlinNativeMojo.includes)
                            setExcludes(this@KotlinNativeMojo.excludes)
                            basedir = sourceBasedir
                            scan()
                            
                            includedFiles.asSequence().map { File(sourceBasedir, it) }
                        }
                    }
                    .map { it.canonicalPath.safeFileName() }
                    .toList()
                
                if (log.isDebugEnabled) {
                    log.debug("Source files:");
                    sources.forEach { log.debug("  $it") }
                }
                log.info("Adding ${sources.size} files to command line arguments");
                addAll(sources);
            }

        } catch (e: MojoExecutionException) {
            throw e;
        } catch (e: RuntimeException) {
            val cause: Throwable? = e.cause;
            throw MojoExecutionException("An error occurred while initializing plugin configuration: ${e.message}", cause);
        } catch (e: Exception) {
            throw MojoExecutionException("An error occurred while initializing plugin configuration: ${e.message}", e);
        }

        super.execute()
    }


    private fun String.safeFileName(): String =
        if (this.contains(" ")) {
            """"$this"""";
        }
        else {
            this;
        }
}

private class FieldAccessor<T>(val target: Any, val field: Field) : Supplier<T?>, Consumer<T?> {

    @Suppress("UNCHECKED_CAST")
    override fun get(): T? {
        return field.get(target) as T;
    }

    fun set(t: T?) {
        field.set(target, t);
    }

    override fun accept(t: T?) {
        set(t)
    }

}